src/main/scala/is/hail/expr/ir/IR.scala

```diff
@@ -56,7 +56,13 @@ final case class AggIn(var typ: TAggregable = null) extends IR
 final case class AggMap(a: IR, name: String, body: IR, var typ: TAggregable = null) extends IR
 final case class AggFilter(a: IR, name: String, body: IR, var typ: TAggregable = null) extends IR
 final case class AggFlatMap(a: IR, name: String, body: IR, var typ: TAggregable = null) extends IR
-final case class AggSum(a: IR, var typ: Type = null) extends IR
+sealed trait ApplyAggOp extends IR {
+  def a: IR
+  def inputType: Type = coerce[TAggregable](a.typ).elementType
+}
+final case class ApplyAggNullaryOp(a: IR, op: NullaryAggOp, var typ: Type = null) extends ApplyAggOp
+final case class ApplyAggUnaryOp(a: IR, op: UnaryAggOp, arg1: IR, var typ: Type = null) extends ApplyAggOp
+final case class ApplyAggTernaryOp(a: IR, op: TernaryAggOp, arg1: IR, arg2: IR, arg3: IR, var typ: Type = null) extends ApplyAggOp
```

Comment:
I had suggested `ApplyAggOp(a: IR, op: AggOp, args: Array[IR], var typ: Type = null) extends IR`.  Do you not like that?

Reply:
I would think that we've implemented the aggops once and now we're done. I think we'll implement lots of Aggregators though and it would be nice to avoid casts in those definitions (makes them less scary to write).

I'll defer to your experience here.

Reply:
Hmm, I think you can take it one step further and get a very nice result: you can write a generic stagedNew that takes the class to create and the `Code`s, and calls the version of newInstance that takes an array of Code[_] for arguments.  (You might need the argument `Class`es, too, but you can get that from the ApplyAggOp types, I think.)  Then you don't need the stagedNew stuff at all when defining a new aggregator, and registering it just amounts to giving the aggregator class and the types coming in and out.

What do you think?